libs/langchain/langchain/vectorstores/redis/redis.py

```diff
@@ -132,84 +99,113 @@ class Redis(VectorStore):
-        metadata_key: str = "metadata",
         vector_key: str = "content_vector",
+        vector_schema: Optional[Dict] = None,
+        metadata_schema: Optional[Dict] = None,
         relevance_score_fn: Optional[Callable[[float], float]] = None,
-        distance_metric: REDIS_DISTANCE_METRICS = "COSINE",
         **kwargs: Any,
     ):
         """Initialize with necessary components."""
-        self.embedding_function = embedding_function
         self.index_name = index_name
+        self.content_key = content_key
+        self.vector_key = vector_key
+        self.embedding_function = embedding_function
         try:
             redis_client = get_client(redis_url=redis_url, **kwargs)
             # check if redis has redisearch module installed
-            _check_redis_module_exist(redis_client, REDIS_REQUIRED_MODULES)
+            check_redis_module_exist(redis_client, REDIS_REQUIRED_MODULES)
         except ValueError as e:
             raise ValueError(f"Redis failed to connect: {e}")
 
+        # Assume import check is done in get_client
+        from langchain.vectorstores.redis.schema import RedisMetadata
+
         self.client = redis_client
-        self.content_key = content_key
-        self.metadata_key = metadata_key
-        self.vector_key = vector_key
-        self.distance_metric = distance_metric
         self.relevance_score_fn = relevance_score_fn
+        self._vector_schema = self._default_content_vector_schema
+        if isinstance(vector_schema, dict):
+            self._vector_schema.update(vector_schema)
+        if metadata_schema:
+            self._metadata_schema = RedisMetadata(**metadata_schema)
+            self._metadata_keys = self._metadata_schema.keys
+        else:
+            self._metadata_schema = None
+            self._metadata_keys = []
 
-    @property
-    def embeddings(self) -> Optional[Embeddings]:
-        # TODO: Accept embedding object directly
-        return None
+        self._distance_metric = self._vector_schema["distance_metric"].upper()
+        self._select_relevance_score_fn()
 
     def _select_relevance_score_fn(self) -> Callable[[float], float]:
         if self.relevance_score_fn:
             return self.relevance_score_fn
 
-        if self.distance_metric == "COSINE":
-            return self._cosine_relevance_score_fn
-        elif self.distance_metric == "IP":
-            return self._max_inner_product_relevance_score_fn
-        elif self.distance_metric == "L2":
-            return self._euclidean_relevance_score_fn
+        metric_map = {
+            "COSINE": self._cosine_relevance_score_fn,
+            "IP": self._max_inner_product_relevance_score_fn,
+            "L2": self._euclidean_relevance_score_fn,
+        }
+        if self._distance_metric in metric_map:
+            return metric_map[self._distance_metric]
         else:
             return _default_relevance_score
 
+    @property
+    def embeddings(self) -> Optional[Embeddings]:
```

Comment:
why is this needed again (ik it's from before, just wondering)? for backwards compatibility?

Reply:
Honestly I don't remember. @tylerhutcherson ?